### PR TITLE
[ENH] Enhance samplers code quality in data module .

### LINKED ISSUE #2155: [ENH] Minor Code Quality Improvements in `GroupedSampler` and `TimeSynchronizedBatchSampler`
## ENH: Minor Code Quality Improvements in `GroupedSampler` and `TimeSynchronizedBatchSampler`

---

## Summary

`GroupedSampler` and `TimeSynchronizedBatchSampler` in `pytorch_forecasting/data/samplers.py`
contain several minor code quality issues that affect readability, developer
experience, and maintainability.

These include:

- a shadowed built-in import (`shuffle`)
- a bare `NotImplementedError` without explanation
- missing return type hints
- a vague warning message
- a small PEP 8 formatting issue
- a missing usage example in a class docstring

This issue proposes addressing these small issues with **non-breaking
improvements** that make the code clearer for contributors and users.

---

## Current Implementation

```python
from sklearn.utils import shuffle

class GroupedSampler(Sampler):

    def __init__(self, sampler, batch_size=64, shuffle=False, drop_last=False):
        # 'shuffle' parameter shadows the imported sklearn shuffle function
        ...

    def get_groups(self, sampler: Sampler):
        # No return type hint, bare NotImplementedError
        raise NotImplementedError()

    def construct_batch_groups(self, groups):
        ...
        warns.append(name)
        if len(warns) > 0:
            warnings.warn(
                f"Less than {self.batch_size} samples available for "
                f"{len(warns)} prediction times. "
                f"Use batch size smaller than {self.batch_size}. "
                f"First 10 prediction times with small batch sizes: {warns[:10]}"
            )

        self._group_index = np.repeat(...)
```

---

### 1. Shadowed Import — `shuffle`

The top-level import

```python
from sklearn.utils import shuffle
```

is shadowed by the `shuffle: bool` parameter in `__init__`.

Inside `__iter__`, the identifier `shuffle` may be interpreted either as the
boolean attribute or the sklearn function, which makes the code harder to read
and could lead to accidental misuse by future contributors.

Example of the ambiguity:

```python
from sklearn.utils import shuffle

def __init__(..., shuffle: bool = False):
    self.shuffle = shuffle

def __iter__(self):
    if self.shuffle:
        groups = {name: shuffle(group) for name, group in self._groups.items()}
```

---

### 2. Bare `NotImplementedError` in `get_groups`

The abstract method `get_groups` raises a bare `NotImplementedError()` with no
message.

```python
def get_groups(self, sampler: Sampler):
    raise NotImplementedError()
```

If a subclass forgets to implement this method, the resulting traceback gives
no indication of what functionality is missing.

---

### 3. Missing Return Type Hint on `get_groups`

The `get_groups` method currently has no return type annotation.

```python
def get_groups(self, sampler: Sampler):
```

This makes it harder for IDEs and static analysis tools to infer the expected
return value.

---

### 4. Vague Warning Message in `construct_batch_groups`

The warning currently instructs the user to reduce the batch size:

```python
warnings.warn(
    f"Less than {self.batch_size} samples available..."
)
```

However, the message does not clearly explain the actual behavior — that these
groups will still be processed but will produce smaller mini-batches.

---

### 5. Missing Blank Line Before Index Construction

A small PEP 8 formatting issue: the comment introducing the index construction
block is not visually separated from the warning block above it.

This slightly reduces readability in the method.

---

### 6. Missing Usage Example in `TimeSynchronizedBatchSampler` Docstring

The `TimeSynchronizedBatchSampler` docstring explains the class behavior but
does not include a simple usage example, which could help new users understand
how to use it.

---

### Fix 1 — Rename sklearn `shuffle` import

Rename the import to avoid shadowing:

```python
from sklearn.utils import shuffle as sklearn_shuffle
```

and update usage:

```python
groups = {name: sklearn_shuffle(group) for name, group in self._groups.items()}
```

---

### Fix 2 — Add descriptive `NotImplementedError`

```python
def get_groups(self, sampler: Sampler) -> dict:
    raise NotImplementedError(
        f"{self.__class__.__name__} must implement `get_groups()` "
        "and return a dictionary mapping group names to index arrays."
    )
```

---

### Fix 3 — Add return type hint to `get_groups`

```python
def get_groups(self, sampler: Sampler) -> dict:
```

---

### Fix 4 — Improve the warning message

```python
warnings.warn(
    f"{len(warns)} prediction time group(s) have fewer than {self.batch_size} samples. "
    f"These groups will yield smaller mini-batches. "
    f"Consider reducing batch_size below {self.batch_size}. "
    f"First 10 affected prediction times: {warns[:10]}"
)
```

---

### Fix 5 — Add blank line before index construction

```python
if len(warns) > 0:
    warnings.warn(...)

# create index from which can be sampled
self._group_index = np.repeat(
    list(self._group_sizes.keys()), list(self._group_sizes.values())
)
```

---

### Fix 6 — Add example to `TimeSynchronizedBatchSampler` docstring

```python
class TimeSynchronizedBatchSampler(GroupedSampler):
    """
    Samples mini-batches randomly but in a time-synchronised manner.

    Example
    -------
    sampler = TimeSynchronizedBatchSampler(base_sampler, batch_size=32, shuffle=True)

    for batch in sampler:
        # batch indices are time-aligned across the mini-batch
        ...
    """
```

---

## Scope

| Item | Details |
|-----|--------|
| File | `pytorch_forecasting/data/samplers.py` |
| Lines changed | ~10–15 |
| Runtime behaviour | Unchanged |
| Breaking changes | None |
| Tests needed | No (logic unchanged, improvements are stylistic and documentation-related) |

---

## Why This Matters

| Improvement | Benefit |
|-------------|--------|
| Rename `shuffle` import | Prevents name shadowing and confusion |
| Descriptive `NotImplementedError` | Helps contributors implementing subclasses |
| Return type hint | Better IDE support and static analysis |
| Improved warning message | Clearer diagnostics for users |
| PEP 8 formatting fix | Better readability |
| Docstring example | Easier onboarding for new users |

---

## Checklist for the PR

- [ ] Rename `shuffle` import to `sklearn_shuffle`
- [ ] Add descriptive message to `NotImplementedError`
- [ ] Add return type hint `-> dict` to `get_groups`
- [ ] Improve warning message in `construct_batch_groups`
- [ ] Add missing blank line for PEP 8 compliance
- [ ] Add usage example to `TimeSynchronizedBatchSampler`
- [ ] Ensure existing tests pass
- [ ] Ensure linting passes (`flake8` / `ruff`)

---

If this improvement sounds useful, I would be happy to open a PR implementing it.

### PR DESCRIPTION
**Pull request by Vishal35198**
# [MNT] Fix shadowed shuffle import and add missing return type hint

#### Reference Issues/PRs

Fixes #2155 

---

#### What does this implement/fix? Explain your changes.

_It focuses on improving the samplers code quality that is present inside the samplers.py inside the data module along with examplers.py , encoders.py ._

This PR addresses two small maintenance issues identified in the codebase:

1. **Shadowed `shuffle` import**
   - The `shuffle` function imported from `sklearn.utils` was shadowed by a local variable.
   - To avoid potential confusion and improve readability, the imported function has been renamed to `sklearn_shuffle`.

2. **Missing return type hint**
   - The `get_groups` method in the `TimeSynchronizedBatchSampler` class (which inherits from `GroupedSampler`) was missing a return type annotation.
   - Added an explicit return type hint using:
     ```python
     from pandas.core.groupby.generic import SeriesGroupBy
     ```
   - The method now correctly specifies the return type as `SeriesGroupBy`.

These changes improve **code clarity, maintainability, and type safety** without modifying runtime behavior.

---

#### What should a reviewer concentrate their feedback on?

- Whether renaming the shadowed `shuffle` import to `sklearn_shuffle` is consistent with the repository’s coding conventions.
- Whether `SeriesGroupBy` is the most appropriate return type annotation for `get_groups`.
- Ensuring the changes maintain compatibility with the existing type expectations of `GroupedSampler`.

---

#### Did you add any tests for the change?

No additional tests were added since:

- The changes are **non-functional improvements** (naming and typing).
- Existing tests pass successfully after the modifications.

---

#### Any other comments?

This PR focuses on **small maintainability improvements** that help reduce ambiguity in the codebase and improve developer experience for contributors working with the sampling utilities.

---

#### PR checklist

- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG].
- [ ] Added/modified tests
- [x] Used pre-commit hooks when committing to ensure that code is compliant with hooks.

### CODE REVIEW

**Comment by Vishal35198** [comment]
@phoeenniixx pls review this PR referencing the issue #2155 .

File: docs/source/tutorials/stallion.ipynb
Comment (phoeenniixx) [inline docs/source/tutorials/stallion.ipynb:1]: please revert these changes
